### PR TITLE
catch errors in setRequestHeader, Fixes #58

### LINKED ISSUE #58: Uncaught xhr exception
I had the following code:
``` purescript
import Prelude
import Control.Monad.Eff (Eff())
import Control.Monad.Eff.Console (CONSOLE(), log)

import Control.Monad.Eff.Exception
import Control.Monad.Aff
import qualified Network.HTTP.Affjax as A
import Network.HTTP.Affjax.Request
import Network.HTTP.Affjax.Response
import Network.HTTP.Method
import Network.HTTP.RequestHeader
import Browser.WebStorage
import Data.Maybe
import Data.Either
import Data.Either.Unsafe

main :: forall e. Eff _ Unit
main = do
  runAff throwException log fetchCurrentAuth

fetchCurrentAuth :: forall eff. Aff _ String
fetchCurrentAuth = do
  token <- liftEff' $ getItem localStorage "Authorization"
  let tok = fromMaybe "" $ fromRight token
  let request = A.defaultRequest
        { url = "url"
        , method = GET
        , headers =
          [ RequestHeader "Content-Type" "application/json"
          , RequestHeader "Accept" "application/json"
          , RequestHeader "Authorization" tok
          ]
        }
  result <- A.affjax request :: A.Affjax _ String
  return $ tok <> result.response
```
The problem ended up being that my token from localStorage had a `\n` on the end of it and when debugging the outputted js I was getting this error `Failed to execute 'setRequestHeader' on 'XMLHttpRequest' […] is not a valid HTTP header field value` that affjax is not catching.

### PR DESCRIPTION
**Pull request by codedmart**

### CODE REVIEW

**Comment by hdgarrood** [comment]
Perhaps we should consider this instead: smart constructors for `RequestHeader`, to ensure that a `RequestHeader` value is always valid?

**Comment by codedmart** [comment]
That is a way better idea. :+1:

**Comment by hdgarrood** [comment]
According to RFC 7230, valid header values may contain:

* visible ascii characters, so I guess that means anything with a char code between 33 and 126 (
(inclusive).
* spaces and tabs
* anything in the "extended ascii" range (128-255).

From: https://tools.ietf.org/html/rfc7230#section-3.2

Additionally, previous versions of HTTP allowed you to have a multiline header value (this mechanism was called "line folding" apparently):

> A sender MUST NOT generate a message that includes
   line folding (i.e., that has any field-value that contains a match to
   the obs-fold rule) unless the message is intended for packaging
   within the message/http media type.

From: https://tools.ietf.org/html/rfc7230#section-3.2.4

I think it's probably safe to assume that affjax won't concern itself with packaging messages within the message/http media type, so I think it should be ok to ban `\r` and `\n` outright.

**Comment by hdgarrood** [comment]
On second thoughts, if we changed `RequestHeader` so that values were guaranteed to be valid, adding this code in too wouldn't hurt. The only thing is, maybe we can make that error message a bit better. Saying "AJAX request failed" sounds like it could be caused by a network issue, to me. Maybe `catch (e) { errback(e) }` would be better, even?

**Comment by garyb** [comment]
I like the suggestion to use the caught error as-is for the `errback` value. :+1: 

There are definitely quite a few things that could be done to improve the construction of requests with more typing, smart constructors, etc. but for now adding the catch in seems like a good compromise, and its presence will do no harm even if we do improve the situation later on.

**Comment by codedmart** [comment]
Fair enough how about now?

**Comment by garyb** [comment]
Is there an advantage in wrapping underlying `e`?

**Comment by codedmart** [comment]
@garyb Are you referring to
```javascript
try {
  xhr.setRequestHeader(header.field, header.value);
}
catch (e) {
  errback(new Error(e))();
}
```
rather then
```javascript
try {
  xhr.setRequestHeader(header.field, header.value);
}
catch (e) errback(new Error(e))();

//or

try xhr.setRequestHeader(header.field, header.value);
catch (e) errback(new Error(e))();
```

**Comment by garyb** [comment]
Oh, sorry I didn't make it clear: I mean using `errback(new Error(e))` rather than just `errback(e)`.

**Comment by codedmart** [comment]
Oh sorry I misunderstood :smile:. No I will make that change in a bit.

**Comment by codedmart** [comment]
OK is that better?

**Comment by garyb** [comment]
Actually, sorry, I've thought of something else here: we probably want the try around the for loop, as otherwise the error handler could be called multiple times if we have a bunch of bad headers. If the error occurs during the construction of the `xhr` object we probably don't want to go ahead and return it either. I'm not too sure what we _do_ want to return when the construction fails. Probably a no-op `Aff` canceller.

**Comment by codedmart** [comment]
@garyb Good call

**Comment by codedmart** [comment]
One of these days I will get it.

**Comment by garyb** [comment]
Looks like it's just the js style checker being picky, the `errback` line is indented two spaces too far apparently.

**Comment by codedmart** [comment]
Sorry I meant to catch that. I use 4 spaces in js by default.

**Comment by codedmart** [comment]
Sorry for all the noise. Let me know if I missed anything.

**Comment by garyb** [comment]
No problem, thanks for the fix!
